packages/daemon/src/services/index.ts

```diff
@@ -289,31 +293,80 @@ export const handleVertexAccepted = async (context: Context, _event: Event) => {
       // for the addresses present on the tx, check if there are any wallets associated
       const addressWalletMap: StringMap<Wallet> = await getAddressWalletInfo(mysql, Object.keys(addressBalanceMap));
 
-      // for each already started wallet, update databases
       const seenWallets = new Set();
-      for (const wallet of Object.values(addressWalletMap)) {
-        const walletId = wallet.walletId;
+      const addressesPerWallet = Object.entries(addressWalletMap).reduce(
+        (result: StringMap<{ addresses: string[], walletDetails: Wallet }>, [address, wallet]: [string, Wallet]) => {
+        const { walletId } = wallet;
 
-        // this map might contain duplicate wallet values, as 2 different addresses might belong to the same wallet
-        if (seenWallets.has(walletId)) continue;
         seenWallets.add(walletId);
-        const { newAddresses, lastUsedAddressIndex } = await generateAddresses(mysql, wallet.xpubkey, wallet.maxGap);
-        // might need to generate new addresses to keep maxGap
-        await addNewAddresses(mysql, walletId, newAddresses, lastUsedAddressIndex);
-        // update existing addresses' walletId and index
+
+        // Initialize the array if the walletId is not yet a key in result
+        if (!result[walletId]) {
+          result[walletId] = {
+            addresses: [],
+            walletDetails: wallet,
+          }
+        }
+
+        // Add the current key to the array
+        result[walletId].addresses.push(address);
+        result[walletId].walletDetails = wallet;
+
+        return result;
+      }, {});
+
+      // Convert to array format expected by getMaxIndicesForWallets
+      const walletDataArray = Object.entries(addressesPerWallet).map(([walletId, data]) => ({
+        walletId,
+        addresses: data.addresses
+      }));
+
+      // Get all max indices in a single query
+      const walletIndices = await getMaxIndicesForWallets(mysql, walletDataArray);
+
+      // Process each wallet
+      for (const [walletId, data] of Object.entries(addressesPerWallet)) {
+        const { walletDetails } = data;
+        const indices = walletIndices.get(walletId);
+
+        if (!indices) {
+          // This is unexpected as we just queried for this wallet
+          logger.error('Failed to get indices for wallet', { walletId });
+          continue;
+        }
+
+        const { maxAmongAddresses, maxWalletIndex } = indices;
+
+        if (!maxAmongAddresses || !maxWalletIndex) {
+          // Do nothing, this is unexpected and an error should have been logged already
+          continue;
+        }
+
+        const diff = maxWalletIndex - maxAmongAddresses;
+
+        if (diff < walletDetails.maxGap) {
+          // We need to generate addresses
+          const addresses = await generateAddresses(walletDetails.xpubkey, maxWalletIndex + 1, walletDetails.maxGap);
```

Comment:
The last parameter is the count of addresses to be generated. You don't need to always generate `maxGap` addresses, it can be `walletDetails.maxGap - diff`, right?

Reply:
Fixed, thanks